src/librustc_lint/builtin.rs

```diff
@@ -1583,3 +1583,46 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ExternCrate {
         self.0 += 1;
     }
 }
+
+declare_lint! {
+    DUPLICATE_ASSOCIATED_TYPE_BINDING,
+    Warn,
+    "warns about duplicate associated type bindings in generics"
+}
+
+#[derive(Copy, Clone)]
+pub struct DupAssocTypeBinding;
+
+impl DupAssocTypeBinding {
+    fn check_bindings(&self, ctxt: &EarlyContext, bindings: &[ast::TypeBinding]) {
+        let mut dup_binds = HashMap::with_capacity(bindings.len());
+        for b in bindings {
+            dup_binds.entry(b.ident.name).or_insert(vec![]).push(b.ident.span);
```

Comment:
So the lint doesn't allocate for every path it checks, I think it'd be better just to do a nested loop search for any duplicates and push their spans to a `Vec`.

Reply:
Your approach is fine, actually, I just think it shouldn't allocate until a duplicate is found, so use `HashMap::new()`. That way we're not allocating when we don't end up using it.

Reply:
Thanks. If we want to reallocate every time, how about keeping the `HashMap` as a member of struct `struct DupAssocTypeBinding`?

Reply:
I don't think that's quite necessary, I was just concerned about allocating every time this function is called.

Reply:
I think you are right. I have updated this function. Thanks.